volmdlr/from_occt.py

```diff
@@ -0,0 +1,328 @@
+
+
+def circle2d_from_occt(curve):
+    """
+    Instanciates a volmdlr Circle2D, from occt object.
+
+    :param curve: OCCT curve.
+    :return: volmdlr Circle2D.
+    """
+    frame = frame2d_from_occt_ax22d(curve.Circ2d().Axis())
+    return curves.Circle2D(frame, curve.Radius())
+
+
+def circle3d_from_occt(curve):
+    """
+    Instanciates a volmdlr Circle3D, from occt object.
+
+    :param curve: OCCT curve.
+    :return: volmdlr Circle3D.
+    """
+    frame = frame3d_from_occt_ax3(curve.Position())
+    return curves.Circle3D(frame, curve.Radius())
+
+
+def ellipse2d_from_occt(curve):
+    """
+    Instanciates a volmdlr Ellipse2D, from occt object.
+
+    :param curve: OCCT curve.
+    :return: volmdlr Ellipse2D.
+    """
+    frame = frame2d_from_occt_ax22d(curve.Position())
+    return curves.Ellipse2D(frame, curve.MajorRadius(), curve.MinorRadius())
+
+
+def ellipse3d_from_occt(curve):
+    """
+    Instanciates a volmdlr Ellipse3D, from occt object.
+
+    :param curve: OCCT curve.
+    :return: volmdlr Ellipse3D.
+    """
+    frame = frame3d_from_occt_ax3(curve.Position())
+    return curves.Ellipse3D(curve.MajorRadius(), curve.MinorRadius(), frame)
+
+
+def hyperbola3d_from_occt(curve):
+    """
+    Instanciates a volmdlr Hyperbola3D, from occt object.
+
+    :param curve: OCCT curve.
+    :return: volmdlr Hyperbola3D.
+    """
+    frame = frame3d_from_occt_ax3(curve.Position())
+    return curves.Hyperbola3D(frame, curve.MajorRadius(), curve.MinorRadius())
+
+
+def parabola3d_from_occt(curve):
+    """
+    Instanciates a volmdlr Parabola3D, from occt object.
+
+    :param curve: OCCT curve.
+    :return: volmdlr Parabola3D.
+    """
+    frame = frame3d_from_occt_ax3(curve.Position())
+    frame = volmdlr.Frame3D(frame.origin, frame.v, frame.u, frame.w)
+    return curves.Parabola3D(frame, curve.Focal())
+
+
+# Edges
+
+
+def bsplinecurve2d_from_occt(curve):
+    """
+    Instanciates a volmdlr BSplineCurve2D, from occt object.
+
+    :param curve: OCCT curve.
+    :return: volmdlr BSplineCurve2D.
+    """
+    control_points = [volmdlr.Point2D(point.X(), point.Y()) for point in curve.Poles()]
+    knots = list(curve.Knots())
+    multiplicities = list(curve.Multiplicities())
+    weigths = None
+    if curve.IsRational():
+        curve.Weights(weights_array := TColStd_Array1OfReal(1, len(control_points)))
+        weigths = list(weights_array)
+    return edges.BSplineCurve2D(curve.Degree(), control_points, multiplicities, knots, weigths)
+
+
+def bsplinecurve3d_from_occt(curve):
+    """
+    Instanciates a volmdlr BSplineCurve3D, from occt object.
+
+    :param curve: OCCT curve.
+    :return: volmdlr BSplineCurve3D.
+    """
+    control_points = [volmdlr.Point3D(point.X(), point.Y(), point.Z()) for point in curve.Poles()]
+    knots = list(curve.Knots())
+    multiplicities = list(curve.Multiplicities())
+    weigths = None
+    if curve.Weights() is not None:
+        weigths = list(curve.Weights())
```

Comment:
Can you test if this works well if you call curve.Weights() a second time after using this method. See: https://github.com/CadQuery/OCP/issues/137 . If it doesn't work i think it's safer to use the same approach as in bsplinecurve2d_from_occt